apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsEmbedControl.tsx

```diff
@@ -0,0 +1,242 @@
+import type { ControlProps, RankedTester } from "@jsonforms/core"
+import {
+  Box,
+  FormControl,
+  HStack,
+  Icon,
+  ListItem,
+  Modal,
+  ModalBody,
+  ModalContent,
+  ModalFooter,
+  ModalHeader,
+  ModalOverlay,
+  Text,
+  UnorderedList,
+  useDisclosure,
+  VStack,
+} from "@chakra-ui/react"
+import { and, isStringControl, rankWith, schemaMatches } from "@jsonforms/core"
+import { withJsonFormsControlProps } from "@jsonforms/react"
+import {
+  Button,
+  FormErrorMessage,
+  FormLabel,
+  Infobox,
+  ModalCloseButton,
+  Textarea,
+} from "@opengovsg/design-system-react"
+import {
+  MAPS_EMBED_URL_REGEXES,
+  VIDEO_EMBED_URL_REGEXES,
+} from "@opengovsg/isomer-components"
+import { BiLink } from "react-icons/bi"
+import { z } from "zod"
+
+import { JSON_FORMS_RANKING } from "~/constants/formBuilder"
+import { useZodForm } from "~/lib/form"
+import {
+  EMBED_NAME_MAPPING,
+  getEmbedNameFromUrl,
+  getIframeSrc,
+} from "../../../utils"
+
+const SUPPORTED_MAPS = Object.keys(MAPS_EMBED_URL_REGEXES).map(
+  (key) => EMBED_NAME_MAPPING[key as keyof typeof MAPS_EMBED_URL_REGEXES],
+)
+
+const SUPPORTED_VIDEOS = Object.keys(VIDEO_EMBED_URL_REGEXES).map(
+  (key) => EMBED_NAME_MAPPING[key as keyof typeof VIDEO_EMBED_URL_REGEXES],
+)
+
+export const jsonFormsEmbedControlTester: RankedTester = rankWith(
+  JSON_FORMS_RANKING.TextControl,
+  and(
+    isStringControl,
+    schemaMatches((schema) => schema.format === "embed"),
+  ),
+)
+
+interface EmbedCodeModalProps {
+  isOpen: boolean
+  onClose: () => void
+  onSave: (embedCode: string) => void
+  urlPattern?: string
+}
+
+function EmbedCodeModal({
```

Comment:
curious - why did we choose to open this in a modal instead of editing with the "label" field on the left-hand sidebar instead? seems like an additional step

Reply:
Hmm it's cos we are only storing the URL in the schema rather than the entire iframe. If we had this directly inside the page editor, then we will either need to reconstruct how the iframe code looks like, or when users paste into the input field, it gets automatically translated into just the URL which might be unexpected.